pkg/meta/base.go

```diff
@@ -1438,6 +1442,66 @@ func (m *baseMeta) CleanupTrashBefore(ctx Context, edge time.Time, increProgress
 	}
 }
 
+func (m *baseMeta) scanTrashFiles(ctx Context, scan trashFileScan) error {
+	var st syscall.Errno
+	var entries []*Entry
+	if st = m.en.doReaddir(ctx, TrashInode, 1, &entries, -1); st != 0 {
+		return errors.Wrap(st, "read trash")
+	}
+
+	type scannedEntry struct {
+		deletedAt time.Time
+		*Entry
+	}
+
+	scannedEntries := make([]scannedEntry, 0, len(entries))
+	for _, e := range entries {
+		ts, err := time.Parse("2006-01-02-15", string(e.Name))
+		if err != nil {
+			logger.Warnf("bad entry as a subTrash: %s", e.Name)
+			continue
+		}
+		scannedEntries = append(scannedEntries, scannedEntry{
+			deletedAt: ts,
+			Entry:     e,
+		})
+	}
+
+	for len(scannedEntries) != 0 {
+		poped := scannedEntries
+		scannedEntries = nil
+		for _, entry := range poped {
+			if entry.Attr.Typ == TypeFile {
+				clean, err := scan(entry.Inode, entry.Attr.Length, entry.deletedAt)
+				if err != nil {
+					return errors.Wrap(err, "scan trash files")
+				}
+				if clean {
+					st = m.en.doUnlink(ctx, entry.Attr.Parent, string(entry.Name))
+					if st != 0 {
+						return errors.Wrap(st, "unlink trash file")
+					}
+				}
+				continue
+			}
+			if entry.Attr.Typ == TypeDirectory {
+				var subEntries []*Entry
+				if st = m.en.doReaddir(ctx, entry.Inode, 1, &subEntries, -1); st != 0 {
+					logger.Warnf("readdir subTrash %d: %s", entry.Inode, st)
+					continue
+				}
+				for _, se := range subEntries {
+					scannedEntries = append(scannedEntries, scannedEntry{
```

Comment:
check whether the entry is file here and increase the accumulate stats if necessary, so we don't have to append files to the list.

Reply:
Here we do a level order traversal, appending directories is necessary.

Reply:
append the se only when it's a directory

Reply:
Refer to https://github.com/juicedata/juicefs/pull/3145#discussion_r1064307684, the directory must be empty, so we do not need level order traversal.

Reply:
Fixed as https://github.com/juicedata/juicefs/pull/3145#discussion_r1065418557